### ISSUE TITLE
Iteration Plan for June 2018 - DRAFT

### ISSUE THREAD

**Issue by kieferrm** [Microsoft/vscode#51483]
This plan captures our work in **June**. This is a 4-week iteration. We will ship our June Update in **early July**.

In May, we previewed several feature: the grid layout for editors, the alternative UI for settings, the file outline in the explorer, and the docked debug toolbar. In June, we continue to focus on those items and try to move them out of their preview state. How far we get there clearly depends on the feedback we hear from you.

This iteration, we also start to officially work on an extension to support pull requests. Dealing with pull requests is a big part of our daily life. Unsurprisingly, we have been exploring ideas in the area for a while but it was too fuzzy where it should go. We now have a more crisp understanding of the approach we want to take. All of our work happens on GitHub and so, with GitHub's help, we'll focus on GitHub support first. The emering APIs will allow for broad vendor support.

## Endgame
- **June 25th, 2018**: Code freeze for the endgame
- **June 29st, 2018**: Endgame done

The endgame details for this iteration are tracked in TBD.

## Plan Items
Below is a summary of the top level plan items.

Legend of annotations:

| Mark | Description |
| ------------- | ------------- |
| :runner: | work in progress |
| :hand: | blocked task |
| :muscle: | stretch goal for this iteration |
| :red_circle: | missing issue reference |
| :large_blue_circle: | more investigation required to remove uncertainty |
| :black_circle: | under discussion within the team |

### Workbench
- [ ] :runner: Finish grid layout for editors #14909 @bpasero @joaomoreno @isidorn
   - [ ] Support centered layout #50936 @bpasero @joaomoreno @isidorn
- [ ] :runner: Continue to improve settings experience #41040 @roblourens
   - [ ] :red_circle: Explore TOC inside the settings editor @roblourens @misolori
- [ ] :runner: Quick box improvements for multi-step inputs #49340 @chrmarti
- [ ] :runner: Provide custom title bar and menus for improved accessibility on Windows #29024 @sbatten
- [ ] :runner: Finish supporting an Outline view (error ticks, select vs. filter option) #5605 @jrieken
- [ ] Explore bread crumb support for navigation #9418 @jrieken
- [ ] :red_circle: Support readonly workspace folders @isidorn
- [ ] :red_circle: Improve rendering of resources contributed by a file system provider @isidorn
- [x] :runner: Support unified way to navigate in the history of input fields (find, change, search, ...) #12645 #50583  @sandy081

### Install
- [ ] Support user level install on Windows #534 @joaomoreno
- [ ] Support portable mode install #329 @joaomoreno

### Terminal
- [ ] Architectural improvements to the renderer and the buffer model @tyriar xtermjs/xterm.js#791

#### Language Server Protocol
- [ ] :red_circle: :runner: Ongoing maintenance and community discusions @dbaeumer
- [ ] Investigate improved visualization of language client and language server interactions octref/lsp-inspector#2 octref/lsp-inspector#3 @octref @dbaeumer
- [ ] :muscle: Explore improved support for in-tool language server tests vscode-language-server-e2e-test#3
vscode-language-server-e2e-test#2 @octref @dbaeumer

#### JavaScript/TypeScript
- [ ] :red_circle: Provide code refactoring in response to rename folder @mjbvz

#### CSS/HTML/JSON
- [ ] Improve support for CSS pseudo classes #41248 @octref
- [ ] Improve accuracy for browser compatibility data of CSS items Microsoft/vscode-css-languageservice#102 @octref
- [ ] :muscle: Gracefully handle unknown at-rules Microsoft/vscode-css-languageservice#51 @octref

### Debug
- [ ] Refined the docked positioning of the debug toolbar #2513 @isidorn @weinand
- [ ] :red_circle: WSL debug support: switch from `bash.exe` to `wsl.exe` @weinand
- [ ] :red_circle: Investigate more fine-grained support for smart-step @roblourens

### Extension Contributions
- [ ] :red_circle: :runner: Support several forthcoming PRs against the `xterm.js` project @tyriar
- [ ] :red_circle: Provide support for pull requests @rmacfarlane @rebornix @shana @daviwil
- [ ] :runner: Improve end-2-end test coverage of the Vetur LS vuejs/vetur#802 @octref

### Extensions
- [ ] :red_circle: Add current extension to workspace recommendations @ramya-rao-a
- [ ] :red_circle: Add ability to freely customize the default Extensions view @ramya-rao-a
- [ ] Provide completion proposals in the search box of the Extensions view #46333 @ramya-rao-a 
- [ ] Allow to dismiss extension recommendations #48743 @jacksonkearl
- [ ] :muscle: In settings search provide a way to show matching but not-installed extensions #49474 @roblourens

### API
- [ ] :red_circle: Allow for case-sensitive/insensitive File System Providers @isidorn
- [ ] :red_circle: Allow for read-only File System Providers @isidorn
- [ ] Support that a completion result can indicate the suggested completion item #35551 @jrieken
- Finalize API for
   - [ ] :runner: Compute definition locations #10037 @mjbvz
   - [ ] :runner: WebView persistence #49022 @mjbvz
   - [ ] :runner: SearchProvider, support to invoke find in files programmatically #47058 @roblourens
   - [ ] :red_circle: :runner: Greying out unused variables @mjbvz
   - [ ] :red_circle: :runner: Subscribing to file rename events @mjbvz
- Propose API for
   - [ ] :runner: Hierachical Document Symbols #34968 @jrieken
   - [ ] :runner: Quick box multi-step inputs #49340 @chrmarti
   - [ ] :red_circle: Adapt WebView API in support for editor grid layout @mjbvz
   - [ ] :red_circle: :runner: Support for custom editor tab icons for webviews @mjbvz
- Custom view API enhancements
   - [ ] :red_circle: Supprt that a custom view can query whether it is visible @sandy081
   - [ ] :red_circle: Support to listen to selection changes @sandy081
   - [ ] :red_circle: Support to execute a command on double click @sandy081

### Engineering
- [ ] :red_circle: Adopt Electron 2.0 @bpasero @tyriar
- [ ] :red_circle: :runner: Improve smoke test reliability @joaomoreno
- [ ] :red_circle: Investigate in tree widget performance @joaomoreno
- [ ] :muscle: Retrain issue classifier bot #49341 @chrmarti
- [ ] :muscle: Convert samples to use tasks 2.0 #50876 @dbaeumer **team**

### Documentation
- [ ] :red_circle: Continue to update our getting started videos @auchenberg

**Comment by rozzzly** [comment]
#### Disenfranchise Community

  [ ] 🏃Continue to ignore user's № 1  requested feature (multi/floating window support) a la #10121 
    (20+ additional (unaddressed) issues available upon request—..._I"m serious, I can link at least 20_)

**Comment by egamma** [comment]
@rozzzly we are not ignoring this feature, but this feature is challenging due to architectural constraints that are explained here: https://github.com/Microsoft/vscode/issues/10121#issuecomment-345497635. 

Also in 1.24 we have added a command that duplicates a workspace in a new window, please see https://code.visualstudio.com/updates/v1_24#_duplicate-workspace-in-new-window.

**Comment by egamma** [comment]
@rozzzly let me step back, as described in our [development process](https://github.com/Microsoft/vscode/wiki/Development-Process) we use a 6 month [roadmap](https://github.com/Microsoft/vscode/wiki/Roadmap) to share what we are planning to work and focus on. We continuously do re-planning as part of the iteration plan. Items that are not on the Roadmap obviously get less attention from our team. Floating windows did not make it into the last Roadmap update in favor of supporting a grid layout for editors. Grid layout was one of the most wanted features at that time and this feature is now available in insiders builds. We update the Roadmap every 6 months and a next update will take place soon. During an update we for sure consider issues with many 👍 votes. We could still decide to not do such a feature due to implementation costs/constraints, this is reality.

Bottom line, the best way to contribute to this process is to 👍 vote for your favorite features. Comments/complaints that we ignore users do not really help much.

**Comment by rozzzly** [comment]
Far enough. Sorry I had to pester you to get a clear response.

**Comment by ayush987goyal** [comment]
I hope #534 doesn't get deferred this time. User level installation is one of the key features most of in corp are waiting for.

**Comment by benfrain** [comment]
Look, I know this isn't necessarily the right place but any chance you could add this to the CSS/HTML/JSON stack; make CSS selectors first class 'symbols' so they can be found in workspaces: https://github.com/Microsoft/vscode/issues/15053

**Comment by j-oliveras** [comment]
@kieferrm I think you want to make a link on the number at `Support centered layout 50936` item.

**Comment by kieferrm** [comment]
@j-oliveras done. Thanks.
